Utils/Subject.h

```diff
@@ -1,104 +1,270 @@
 #pragma once
 
-#include "Observer.h"
-
 #include <vector>
 #include <functional>
+#include <memory>
+
+#include "ReflectionUtils.h"
 
 namespace leap
 {
 	class Subject
 	{
-	public:
-		Subject() = default;
-		~Subject()
+	private:
+		struct MemberFunctionWrapper
 		{
-			for (auto& observer : m_Observers)
+			virtual ~MemberFunctionWrapper() = default;
+			virtual void Call() const = 0;
+			virtual unsigned int GetId() const = 0;
+			virtual bool IsConst() const = 0;
+		};
+
+		template<typename T>
+		struct ImplMemberFunctionWrapper final : public MemberFunctionWrapper
+		{
+			ImplMemberFunctionWrapper(T* _pThis, void(T::* _pFunction)())
+				: pThis{ _pThis }, pFunction{ _pFunction }, id{ ReflectionUtils::GenerateTypenameHash<T>() }
```

Comment:
Initialization order will not be executed like this (same for Const class type).
id,
pThis,
pFunctuin